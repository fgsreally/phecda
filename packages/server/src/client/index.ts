import type { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios'
import type { MergeType, RequestType, ResOrErr, Transform } from '../types'
interface RequestArgs {
  body: Record<string, any>
  query: Record<string, string>
  params: Record<string, string>
  realParam: string
  method: RequestType
  url: string
  name: string
}
type MergedReqArg = Pick<RequestArgs, 'body' | 'query' | 'params'>
export function toReq(arg: RequestArgs) {
  const { body, query, realParam, method, url } = arg
  return { method, url, body, query: Object.keys(query).length > 0 ? `?${Object.entries(query).map(([k, v]) => `${k}=${v}`).join('&')}` : '', params: realParam }
}
// @ts-expect-error misdirction
export const merge: MergeType = (...args: RequestArgs[]) => {
  const ret = {} as Record<string, MergedReqArg>
  for (const i of args) {
    const { name, body, query, params } = i
    ret[name] = { body, query, params }
  }

  return ret
}
export type RequestMethod = <F extends (...args: any[]) => any >(fn: F, args: Parameters<F>) => Promise<ReturnType<F>>

export function createReq(instance: AxiosInstance): <R>(arg: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<Awaited<R>> > {
  // @ts-expect-error methods without route decorator won't send request
  return (arg: any, config?: AxiosRequestConfig) => {
    const { url, params, query, body, method } = toReq(arg as RequestArgs)
    if (!method) {
      console.warn('methods without route decorator won\'t send request')
      return
    }

    const ret = [`${url}${params}${query}`] as any[]
    body && ret.push(body)
    config && ret.push(config)
    // @ts-expect-error misdirction

    return instance[method](...ret)
  }
}

export function createMergeReq(instance: AxiosInstance, key = '/__PHECDA_SERVER__'): < R extends any[]>(args: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<ResOrErr<R>>> {
  // @ts-expect-error misdirction

  return (args: Record<string, MergedReqArg>, config?: AxiosRequestConfig) => {
    return instance.post(key, args, config)
  }
}

export function P<C extends new (...args: any) => any>(Model: C): Transform<C> {
  return new Model()
}
