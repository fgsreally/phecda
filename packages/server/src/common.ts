export const MERGE_SYMBOL = '__PS_MERGE__'
export const UNMOUNT_SYMBOL = '__PS_UNMOUNT__'
export const MODULE_SYMBOL = '__PS_MODULE__'
export const META_SYMBOL = '__PS_META__'
export const APP_SYMBOL = '__PS__'
export const IS_DEV=process.env.NODE_ENV === 'development'