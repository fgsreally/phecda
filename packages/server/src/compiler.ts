import type { RequestType } from './types'

export class Pcompiler {
  content = ''
  classMap: Record<string, { [key: string]: string }> = {}
  constructor() { }

  getContent() {
    let content = ''

    for (const name in this.classMap) {
      content += `
        export class ${name}{
            ${Object.values(this.classMap[name]).reduce((p, c) => p + c)}
            }`
    }
    return content
  }

  addMethod(className: string, methodName: string, route = '', requestType: RequestType | '' = '', params: { type: string; key: string; index: number }[] = []) {
    const url = route.replace(/\/\:([^\/]*)/g, '')
    if (!this.classMap[className])
      this.classMap[className] = {}
    this.classMap[className][methodName] = `
    ${methodName}(${genParams(params)}){
const ret={name:"${className}-${methodName}",body:{},query:{},params:{},realParam:'',method:"${requestType}",url:"${url}"}
${params.reduce((p, c) => `${p}ret.${c.type}.${c.key}=arg${c.index}\n${c.type === 'params' ? `ret.realParam+='/'+arg${c.index}\n` : ''}`, '')}
return ret
    }
    `
  }
}

// function genExpression(type: string, key: string, index: number) {
//   switch (type) {
//     case 'params':
//       return `ret.params+='/'+arg${index}`
//     case 'query':
//       return `ret.query+=ret.query?"&${key}="+arg${index}:"?${key}="+arg${index}`
//     case 'body':
//       return `ret.body[${key}]=arg${index}`
//   }
// }

function genParams(decorators: any[]) {
  let index = 0
  return decorators.reduce((p) => {
    return `${`${p}arg${index++}`},`
  }, '')
}
