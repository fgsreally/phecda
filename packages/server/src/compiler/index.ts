export * from './rpc'
export * from './http'
