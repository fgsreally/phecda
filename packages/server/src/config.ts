export const Pconfig = {
  rabbitmq: {
    guard: false,
    interceptor: false,
  },
}
