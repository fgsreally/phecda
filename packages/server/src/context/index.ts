export * from './server'
export * from './base'
export * from './micro'
