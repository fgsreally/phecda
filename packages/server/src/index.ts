import 'reflect-metadata'

type Construct<T = any> = new (...args: Array<any>) => T

export function Factory<T>(Module: Construct<T>) {
  const moduleMap = new Map<Construct, InstanceType<Construct>>()
  const app = buildNestModule(Module, moduleMap) as InstanceType<Construct<T>>
  return { app, moduleMap }
}

function buildNestModule(Module: Construct, map: Map<Construct, InstanceType<Construct>>) {
  const paramtypes = getParamtypes(Module) as Construct[]
  let instance: InstanceType<Construct>
  if (map.has(Module)) {
    instance = map.get(Module)
    if (!instance)
      throw new Error(`exist Circular Module dep--${Module}`)
    return instance
  }
  map.set(Module, undefined)
  if (paramtypes) {
    instance = new Module(...paramtypes.map(item =>
      buildNestModule(item, map),
    ))
  }
  else {
    instance = new Module()
  }
  map.set(Module, instance)

  return instance
}

function getParamtypes(Module: Construct) {
  return Reflect.getMetadata('design:paramtypes', Module)
}
