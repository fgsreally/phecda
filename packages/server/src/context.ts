import pc from 'picocolors'
import { defaultPipe } from './pipe'
import { ForbiddenException, FrameworkException } from './exception'
import { defaultFilter } from './filter'
import { Histroy } from './history'
import type { P } from './types'
import { IS_DEV, IS_STRICT } from './common'
import type { Meta } from './meta'
import { log } from './utils'
import type { Exception } from './exception'

export type GuardType<C extends P.BaseContext = any> = ((ctx: C) => Promise<boolean> | boolean)
export type InterceptorType<C extends P.BaseContext = any> = (ctx: C) => (any | ((ret: any) => any))
export type PipeType<C extends P.BaseContext = any> = (arg: { arg: any; option?: any; key: string; type: string; index: number; reflect: any }, ctx: C) => Promise<any>
export type FilterType<C extends P.BaseContext = any, E extends Exception = any> = (err: E | Error, ctx?: C) => Error | any

export class Context<Data extends P.BaseContext> {
  method: string
  params: string[]
  history = new Histroy()

  static filterRecord: Record<PropertyKey, FilterType> = {
    default: defaultFilter,
  }

  static pipeRecord: Record<PropertyKey, PipeType> = {
    default: defaultPipe,
  }

  static guardRecord: Record<PropertyKey, GuardType> = {}
  static interceptorRecord: Record<PropertyKey, InterceptorType> = {}

  static pluginRecord: Record<PropertyKey, any> = {}
  postInterceptors: Function[]

  constructor(public data: Data) {
    if (IS_DEV)
      // @ts-expect-error work for debug
      data._context = this
  }

  usePipe(args: { arg: any; pipe?: string; pipeOpts?: any; type: string; key: string; index: number; reflect: any }[]) {
    return Promise.all(args.map((item) => {
      if (item.pipe && !Context.pipeRecord[item.pipe]) {
        if (IS_STRICT)
          throw new FrameworkException(`can't find pipe named '${item.pipe}'`)

        else
          return Context.pipeRecord.default(item, this.data)
      }

      return Context.pipeRecord[item.pipe || 'default'](item, this.data)
    }))
  }

  useFilter(arg: any, filter = 'default') {
    if (!Context.filterRecord[filter]) {
      if (IS_STRICT)
        throw new FrameworkException(`can't find filter named '${filter}'`)
      else
        return Context.filterRecord.default(arg, this.data)
    }

    return Context.filterRecord[filter](arg, this.data)
  }

  async useGuard(guards: string[]) {
    for (const guard of guards) {
      if (this.history.record(guard, 'guard')) {
        if (!(guard in Context.guardRecord)) {
          if (IS_STRICT)
            throw new FrameworkException(`can't find guard named '${guard}'`)
          continue
        }
        if (!await Context.guardRecord[guard](this.data))
          throw new ForbiddenException(`Guard exception--${guard}`)
      }
    }
  }

  async usePostInterceptor(ret: any) {
    for (const cb of this.postInterceptors)
      ret = await cb(ret) || ret

    return ret
  }

  async useInterceptor(interceptors: string[]) {
    const ret = []
    for (const interceptor of interceptors) {
      if (this.history.record(interceptor, 'interceptor')) {
        if (!(interceptor in Context.interceptorRecord)) {
          if (IS_STRICT)
            throw new FrameworkException(`can't find interceptor named '${interceptor}'`)

          continue
        }
        const postInterceptor = await Context.interceptorRecord[interceptor](this.data)
        if (postInterceptor !== undefined) {
          if (typeof postInterceptor === 'function')
            ret.push(postInterceptor)

          else
            return postInterceptor
        }
      }
    }
    this.postInterceptors = ret
  }

  static usePlugin(plugins: string[]) {
    const ret = []
    for (const m of plugins) {
      if (!(m in Context.pluginRecord)) {
        if (IS_STRICT)
          throw new FrameworkException(`can't find middleware named '${m}'`)

        continue
      }
      ret.push(Context.pluginRecord[m])
    }
    return ret as any[]
  }
}

export function addPlugin<T>(key: PropertyKey, handler: T) {
  if (Context.pluginRecord[key] && Context.pluginRecord[key] !== handler)
    log(`overwrite Plugin "${String(key)}"`, 'warn')

  Context.pluginRecord[key] = handler
}

export function addPipe<C extends P.BaseContext>(key: PropertyKey, handler: PipeType<C>) {
  if (Context.pipeRecord[key] && Context.pipeRecord[key] !== handler)
    log(`overwrite PipeType "${String(key)}"`, 'warn')
  Context.pipeRecord[key] = handler
}

export function addFilter<C extends P.BaseContext>(key: PropertyKey, handler: FilterType<C>) {
  if (Context.filterRecord[key] && Context.filterRecord[key] !== handler)
    log(`overwrite FilterType "${String(key)}"`, 'warn')
  Context.filterRecord[key] = handler
}

export function addGuard<C extends P.BaseContext>(key: PropertyKey, handler: GuardType<C>) {
  if (Context.guardRecord[key] && Context.guardRecord[key] !== handler)
    log(`overwrite GuardType "${String(key)}"`, 'warn')
  Context.guardRecord[key] = handler
}

export function addInterceptor<C extends P.BaseContext>(key: PropertyKey, handler: InterceptorType<C>) {
  if (Context.interceptorRecord[key] && Context.interceptorRecord[key] !== handler)
    log(`overwrite InterceptorType "${String(key)}"`, 'warn')
  Context.interceptorRecord[key] = handler
}

// detect whether plugin/filter/pipe/guard/intercept is injected
export function isAopDepInject(meta: Meta[], { guards, interceptors, plugins }: {
  guards?: string[]
  interceptors?: string[]
  plugins?: string[]
} = {}) {
  const pluginSet = new Set<string>(plugins)

  const guardSet = new Set<string>(guards)
  const interceptorSet = new Set<string>(interceptors)
  const pipeSet = new Set<string>()

  const filterSet = new Set<string>()
  meta.forEach(({ data }) => {
    if (data.filter)
      filterSet.add(data.filter)

    data.interceptors.forEach(i => interceptorSet.add(i))
    data.guards.forEach(i => guardSet.add(i))
    data.plugins.forEach(i => pluginSet.add(i))
    data.params.forEach((i) => {
      if (i.pipe)
        pipeSet.add(i.pipe)
    })
  })

  const missPlugins = [...pluginSet].filter(i => !Context.pluginRecord[i])
  const missGuards = [...guardSet].filter(i => !Context.guardRecord[i])
  const missInterceptors = [...interceptorSet].filter(i => !Context.interceptorRecord[i])
  const missPipes = [...pipeSet].filter(i => !Context.pipeRecord[i])
  const missFilters = [...filterSet].filter(i => !Context.filterRecord[i])

  if (missPlugins.length)
    log(`${pc.white(`Plugin [${missPlugins.join(',')}]`)} doesn't exist`, 'warn')
  if (missGuards.length)
    log(`${pc.magenta(`GuardType [${missGuards.join(',')}]`)} doesn't exist`, 'warn')

  if (missInterceptors.length)
    log(`${pc.cyan(`InterceptorType [${missInterceptors.join(',')}]`)} doesn't exist`, 'warn')

  if (missPipes.length)
    log(`${pc.blue(`PipeType [${missPipes.join(',')}]`)} doesn't exist`, 'warn')

  if (missFilters.length)
    log(`${pc.red(`FilterType [${missFilters.join(',')}]`)} doesn't exist`, 'warn')
}
