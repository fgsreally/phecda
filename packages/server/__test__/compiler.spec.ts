import { describe, expect, it } from 'vitest'
import { Body, Controller, Factory, Get, Query, generateHTTPCode, generateRPCCode } from '../src'

describe('compiler generate code', () => {
  it('generate http request', async () => {
    @Controller('/base')
    class A {
      @Get('/test')
      test(@Query('id') id: string, @Body('name') name: string) {
        return id + name
      }
    }
    const { meta } = await Factory([A])
    const code = generateHTTPCode(meta.map(item => item.data))

    expect(code).toMatchSnapshot()
  })

  it('generate RPC request', async () => {
    @Controller('/base')
    class A {
      @Get('/test')
      test(@Body() id: string, @Body() name: string) {
        return id + name
      }
    }
    const { meta } = await Factory([A])
    const code = generateRPCCode(meta.map(item => item.data))
    expect(code).toMatchSnapshot()
  })
})
