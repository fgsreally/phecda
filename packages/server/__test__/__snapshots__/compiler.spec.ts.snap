// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler generate code > generate RPC request 1`] = `""`;

exports[`compiler generate code > generate http request 1`] = `
"
        export class A{
            
    test(...args){
const ret={tag:"A-test",body:{},headers:{},query:{},params:{},method:"get",url:"/base/test",args}

ret.query['id']=args[0]

ret.body['name']=args[1]


return ret
    }
    
            }"
`;
